WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteStateProvider.kt

```diff
@@ -0,0 +1,72 @@
+package org.wordpress.android.ui.mysite
+
+import androidx.lifecycle.asFlow
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.flatMapMerge
+import kotlinx.coroutines.flow.flattenMerge
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.withContext
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.ui.mysite.MySiteUiState.PartialState.SelectedSite
+import org.wordpress.android.ui.mysite.MySiteUiState.PartialState.ShowSiteIconProgressBar
+
+class MySiteStateProvider(
+    private val mainDispatcher: CoroutineDispatcher,
+    private val selectedSiteRepository: SelectedSiteRepository,
+    vararg sources: MySiteSource<*>
+) {
+    private val mySiteSources = mutableListOf<MySiteSource<*>>(
+            selectedSiteSource(),
+            siteIconProgressSource()
+    ).apply {
+        addAll(sources)
+    }
+
+    @FlowPreview
+    val state: Flow<MySiteUiState> = selectedSiteRepository.siteSelected.asFlow().flatMapMerge { siteId ->
+        if (siteId != null) {
+            mySiteSources.map { it.buildSource(siteId).distinctUntilChanged() }.asFlow().flattenMerge()
+        } else {
+            flowOf()
+        }
+    }.let { partialStates ->
+                flow {
+                    var accumulator = MySiteUiState()
+                    emit(accumulator)
```

Comment:
the idea here (and maybe it's unnecessary) is to clear the data when you switch site before anything else is loaded. When I'm thinking about it, it might not be necessary because the "SiteModel" is emitted right after so it shouldn't be necessary to emit the empty model.